pkg/fuzzer/queue/queue.go

```diff
@@ -0,0 +1,375 @@
+		ret := pq.queue[pq.pos]
+		pq.queue[pq.pos] = nil
+		pq.pos++
+		if pq.stat != nil {
+			pq.stat.Add(-1)
+		}
+		return ret
+	}
+	return nil
+}
+
+// SourceMultiplexer combines several different sources in a particular order.
+// It is NOT thread-safe.
+type SourceMultiplexer struct {
+	sources []Source
+}
+
+func Multiplex(sources ...Source) Source {
+	return &SourceMultiplexer{sources: sources}
+}
+
+func (sm *SourceMultiplexer) Next() *Request {
+	for _, s := range sm.sources {
+		req := s.Next()
+		if req != nil {
+			return req
+		}
+	}
+	return nil
+}
+
+type callback struct {
+	cb func() *Request
+}
+
+// Callback produces a source that calls the callback to serve every Next() request.
+func Callback(cb func() *Request) Source {
+	return &callback{cb}
+}
+
+func (cb *callback) Next() *Request {
+	return cb.cb()
+}
+
+type alternate struct {
+	base Source
+	mu   sync.Mutex
+	rnd  *rand.Rand
+	prob float32
+}
+
+// Alternate returns a nil *Request in `prob` share of requests.
+func Alternate(base Source, rnd *rand.Rand, prob float32) Source {
+	return &alternate{
+		base: base,
+		rnd:  rnd,
+		prob: prob,
+	}
+}
+
+func (a *alternate) Next() *Request {
+	var skip bool
+	if a.mu.TryLock() {
+		skip = a.rnd.Float32() < a.prob
+		a.mu.Unlock()
+	}
+	if skip {
+		return nil
+	}
+	return a.base.Next()
+}
+
+type PriorityQueue struct {
+	mu       *sync.Mutex
+	ops      *priorityQueueOps[*Request]
+	currPrio priority
+}
+
+func Priority() *PriorityQueue {
+	return &PriorityQueue{
+		mu:       &sync.Mutex{},
+		ops:      &priorityQueueOps[*Request]{},
+		currPrio: priority{0},
+	}
+}
+
+func (pq *PriorityQueue) AppendQueue() *PriorityQueue {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	nextPrio := append(priority{}, pq.currPrio...)
+	return &PriorityQueue{
+		// We use the same queue, therefore the same mutex.
+		mu:       pq.mu,
+		ops:      pq.ops,
+		currPrio: append(nextPrio, 0),
+	}
+}
+
+func (pq *PriorityQueue) Submit(req *Request) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	pq.ops.Push(req, pq.currPrio)
+}
+
+func (pq *PriorityQueue) Next() *Request {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.ops.Pop()
+}
+
+type DynamicMultiplexer struct {
+	mu   sync.Mutex
+	list []Source
+}
+
+func (dm *DynamicMultiplexer) Add(source Source) {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.list = append(dm.list, source)
+}
+
+func (dm *DynamicMultiplexer) Next() *Request {
+	dm.mu.Lock()
+	list := dm.list
+	dm.mu.Unlock()
+	for _, source := range list {
+		req := source.Next()
+		if req != nil {
+			return req
+		}
+	}
+	return nil
+}
+
+// Deduplicator() keeps track of the previously run requests to avoid re-running them.
+type Deduplicator struct {
+	mu     sync.Mutex
+	ctx    context.Context
+	source Source
+	hash   func(*Request) hash.Sig
+	mm     map[hash.Sig]*Request
+}
+
+func NewDeduplicator(ctx context.Context, source Source, hashFunc func(*Request) hash.Sig) *Deduplicator {
```

Comment:
I think this hashFunc should be part of Deduplicator rather than passed to ctor.
The intention of my hash func in vminfo was to hash all of Request, maybe I missed some fields but that was just a short cut b/c I knew what fields are used/matter.

Reply:
Are you aware of any standard ways to hash an arbitrary complex struct in Go? There are many fields in Request, so I'm not sure we want to list all of them in the hash function. 

The GoB serializer should theoretically work well as long as we don't use maps, where I assume the serialization must differ each time.

Reply:
Not sure.
But if we test that duplicates are actually equal, then even fragile hash func may be fine (tests will fail as soon as it become wrong). Maybe if we do reflect.DeepEqual for both requests and panic if they are not equal.